File: aoc23/day1/main.py
from typing import Dict, List, Optional

digit_representations: Dict[str, int] = {
    "1": 1, "2": 2, "3": 3, "4": 4, "5": 5, "6": 6, "7": 7, "8": 8, "9": 9,
    "one": 1, "two": 2, "three": 3, "four": 4, "five": 5,
    "six": 6, "seven": 7, "eight": 8, "nine": 9,
}


def digit(s: str) -> Optional[int]:
    return next((d for t, d in digit_representations.items() if s.endswith(t)), None)


def digits(s: str) -> List[int]:
    return list(
        filter(lambda x: x is not None,
               map(lambda substr: digit(substr),
                   (s[0:end] for end in range(len(s))))))


def value(s: str) -> int:
    d = digits(s)
    return d[0] * 10 + d[-1]


def run():
    with open("input.txt") as f:
        print(sum([value(s) for s in f.readlines()]))


run()

File: aoc23/day1/src/main.rs
#[macro_use]
extern crate lazy_static;

use num2words::{Lang, Num2Words};
use std::collections::HashMap;
use std::fs;

lazy_static! {
    static ref DIGITS: HashMap<String, u32> = {
        let mut map = HashMap::new();
        for i in 1..=9 {
            map.insert(i.to_string(), i);
            map.insert(Num2Words::new(i).lang(Lang::English).to_words().unwrap(), i);
        }
        map
    };
}

fn value_two(line: &str) -> u32 {
    let (first, last) = (0..line.len())
        .filter_map(|start| {
            DIGITS
                .iter()
                .flat_map(|(s, d)| line[start..].starts_with(s).then_some(d))
                .next()
        })
        .copied()
        .fold((None, None), |(f, _), x| (f.or(Some(x)), Some(x)));
    first.unwrap() * 10 + last.unwrap()
}

fn value_one(line: &str) -> u32 {
    let (first, last) = line
        .chars()
        .filter_map(|c| c.to_digit(10))
        .fold((None, None), |(f, _), x| (f.or(Some(x)), Some(x)));
    first.unwrap() * 10 + last.unwrap()
}

fn part_two(lines: &[&str]) -> u32 {
    lines.iter().map(|line| value_two(line)).sum()
}

fn part_one(lines: &[&str]) -> u32 {
    lines.iter().map(|line| value_one(line)).sum()
}

fn main() {
    let input = fs::read_to_string("inputs/day1.txt").unwrap();
    let lines = input
        .split('\n')
        .filter(|s| !s.is_empty())
        .collect::<Vec<_>>();
    println!("Part one: {}", part_one(&lines));
    println!("Part two: {}", part_two(&lines));
}
